flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IndividualRestartsConcurrencyTest.java

```diff
@@ -83,6 +84,13 @@
  */
 public class IndividualRestartsConcurrencyTest extends TestLogger {
```

Comment:
In our current implementation of the `RestartIndividualStrategy` we have no longer concurrency. It is a bit odd that we only introduce it for this test case via the `testExecutor`. Moreover, this test is in my opinion more a `ConcurrentFailoverStrategyExecutionGraphTest` which tests that the `ExecutionGraph` properly restricts state changes wrt the `globalVersionMod` which is not necessarily bound restricted to the `RestartIndividualStrategy`. What I would suggest is the following: Remove the `testExecutor` from the `RestartIndividualStrategy` and introducing a `ManuallyTriggeredFailoverStrategy`. The `ManuallyTriggeredFailoverStrategy` is only for testing purposes and might simply extend `RestartIndividualStrategy`. The `ManuallyTriggeredFailoverStrategy` will simply collect all calls to the `restartProcedure` of the `RestartIndividualStrategy` which for this purpose should be factored out into its own overridable method. Then we can leave the tests basically as they are using the `TestComponentMainThreadExecutor#forMainThread`.

Reply:
We should however keep. `testLocalFailureFailsPendingCheckpoints`.

Reply:
The concurrent tests are actually covered by the `PipelinedRegionFailoverConcurrencyTest`. By controlling when an `Execution` is cancelled we can effectively control when the restart behaviour is triggered on the `ExecutionGraph`. That way, we don't need to add custom concurrent `FailoverStrategies` just for testing purposes.